### ISSUE TITLE
[clangd] Incorrect navigation caused by deducing this

### ISSUE THREAD

**Issue by ymyh** [llvm/llvm-project#116335]
```c++
struct B
{
    template<typename Self>
    auto bar(this Self &&) 
    {
        return 0;
    }

    auto bar2(this B &&)
    {
        return 0;
    }
};

struct A
{
    auto foo()
    {
        return B{};
    }
};

auto baz(int n)
{
    return n;
}

int main()
{
    A a{};
    baz(a.foo().bar());

    A a2{};
    int n  = a2.foo().bar2();
}
```
incorrect navigation of a.foo and a to baz, a2.foo and a2 to n

**Comment by HighCommander4** [comment]
Slightly reduced:

```c++
struct S
{
    int bar(this S&);
};

int main()
{
    S waldo;
    int x = waldo.bar();
}
```

Go-to-def on `waldo` in `waldo.bar()` incorrectly goes to `x`.

**Comment by HighCommander4** [comment]
`SelectionTree` is incorrectly selecting the `x`.

Here is a SelectionTree trace:

```
Computing selection for </clangd-test/TestTU.cpp:7:19, col:24>
 skip: CXXRecordDecl </clangd-test/TestTU.cpp:2:9, line:4:9>
 push: FunctionDecl </clangd-test/TestTU.cpp:5:9, line:8:9>
  push: NestedNameSpecifierLoc <<invalid sloc>>
  pop: NestedNameSpecifierLoc
  skip: FunctionProtoTypeLoc </clangd-test/TestTU.cpp:5:9, col:18>
  push: CompoundStmt </clangd-test/TestTU.cpp:5:20, line:8:9>
   skip: DeclStmt </clangd-test/TestTU.cpp:6:11, col:18>
   push: DeclStmt </clangd-test/TestTU.cpp:7:11, col:30>
    push: VarDecl </clangd-test/TestTU.cpp:7:11, col:29>
     push: NestedNameSpecifierLoc <<invalid sloc>>
     pop: NestedNameSpecifierLoc
     skip: BuiltinTypeLoc </clangd-test/TestTU.cpp:7:11>
     skip: CallExpr </clangd-test/TestTU.cpp:7:25, col:29>
    pop: VarDecl
     hit selection: </clangd-test/TestTU.cpp:7:11, col:29>
   pop: DeclStmt
  pop: CompoundStmt
 pop: FunctionDecl
Built selection tree
 TranslationUnitDecl 
   FunctionDecl void foo()
     CompoundStmt { …
       DeclStmt int x = bar(waldo);
        .VarDecl int x = bar(waldo)
```

**Comment by HighCommander4** [comment]
Basically, the hit-testing algorithm skips the `CallExpr` node for `waldo.bar()` altogether, because `CallExpr::getBeginLoc()` returns the callee's begin loc, which is the `bar` token, so we think the `waldo` token falls outside of the `CallExpr`.

**Comment by HighCommander4** [comment]
> `CallExpr::getBeginLoc()` returns the callee's begin loc, which is the `bar` token

Let's start by seeing if we can get this fixed upstream in the AST: https://github.com/llvm/llvm-project/issues/116335

**Comment by HighCommander4** [comment]
> > `CallExpr::getBeginLoc()` returns the callee's begin loc, which is the `bar` token
> 
> Let's start by seeing if we can get this fixed upstream in the AST: [#116928](https://github.com/llvm/llvm-project/issues/116928)

This was fixed upstream in #117841. I also included a clangd test in that patch, so there is nothing further to do here.
